server/apps/api_notify.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package apps
+
+import (
+	"github.com/mattermost/mattermost-plugin-apps/server/constants"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type SubscriptionNotification struct {
+	SubscriptionID SubscriptionID
+	Subject        constants.SubscriptionSubject
+	ChannelID      string
+	ParentID       string
+	PostID         string
+	RootID         string
+	TeamID         string
+	UserID         string
+	Expanded       *Expanded
+}
+
+// Notify sends a POST change notification for a set of subscriptions
+func (s *Service) Notify(subject constants.SubscriptionSubject, tm *model.TeamMember, cm *model.ChannelMember, actingUser *model.User, channel *model.Channel, post *model.Post) error {
+	actingUserID := ""
+	channelOrTeamID := ""
+
+	msg := &SubscriptionNotification{
+		Subject: subject,
+		UserID:  actingUserID,
+	}
+
+	if actingUser != nil && actingUser.Id != "" {
+		actingUserID = actingUser.Id
+	}
+	if channel != nil {
+		msg.ChannelID = channel.Id
+	}
+	if cm != nil {
+		msg.ChannelID = cm.ChannelId
+		channelOrTeamID = cm.ChannelId
+	}
+	if tm != nil {
+		msg.TeamID = tm.TeamId
+		msg.UserID = tm.UserId
+		channelOrTeamID = tm.TeamId
+	}
+	if post != nil {
+		msg.PostID = post.Id
+		msg.ParentID = post.ParentId
+		msg.RootID = post.RootId
+	}
```

Comment:
FYI, I hate this code, but wanted to get this PR to you.  This code determines how to get IDs for the `Expand()` functions.  

I'm not sure of the difference between actingUser and User.  It's very possible I'm not handling that correctly.  See line 30 where I set UserID: actingUserID

Reply:
Right. In the case when no obvious "actingUser" is available, let's set it to nil.

Reply:
`UserID` should be the subject of the action, aka UserJoinedChannel - would be the userID that joined. For a Post, it's the UserId on the Post. It should rarely (if ever?) be set to actingUserId

Reply:
@lev, maybe the API docs are what confused me (and why I had set `userID` to `actingUserID`.  Looking at [Plugin server reference](https://developers.mattermost.com/extend/plugins/server/reference/) all the `UserHas...` hooks have the `model.User` in the function signature set named as `actor`


![image](https://user-images.githubusercontent.com/7575921/95384172-d6bee380-08b1-11eb-900a-cae25e5687cc.png)

Reply:
Is the actor not the user that left/joined each associated team/channel?